crates/sui-types/src/nitro_attestation.rs

```diff
@@ -391,138 +400,215 @@ impl AttestationDocument {
     /// Parse the payload of the attestation document, validate the cert based on timestamp, and the pcrs match.
     /// Adapted from <https://github.com/EternisAI/remote-attestation-verifier/blob/main/src/lib.rs>
     pub fn parse_payload(
-        payload: &Vec<u8>,
+        payload: &[u8],
     ) -> Result<AttestationDocument, NitroAttestationVerifyError> {
-        let document_data: ciborium::value::Value = ciborium::de::from_reader(payload.as_slice())
-            .map_err(|err| {
-            NitroAttestationVerifyError::InvalidAttestationDoc(format!(
-                "cannot parse payload CBOR: {}",
-                err
-            ))
-        })?;
+        let document_map = Self::to_map(payload)?;
+        Self::validate_document_map(&document_map)
+    }
+
+    fn to_map(payload: &[u8]) -> Result<HashMap<String, Value>, NitroAttestationVerifyError> {
+        let document_data: ciborium::value::Value =
+            ciborium::de::from_reader(payload).map_err(|err| {
+                NitroAttestationVerifyError::InvalidAttestationDoc(format!(
+                    "cannot parse payload CBOR: {}",
+                    err
+                ))
+            })?;
 
-        let document_map = match document_data {
-            ciborium::value::Value::Map(map) => map,
+        let document_map: HashMap<String, Value> = match document_data {
```

Comment:
is there a reasonable length limit on payload we can use?

Reply:
its not documented anywhere, although we can provide a conservative estimation (we have a limit for most of the fields, but after parsing). alternatively, we can rely on the max txn input limit for that.

Reply:
if parsing is cheap, then depending on tx limit makes sense; if it doesn't, we might want to limit it here
(e.g., if parsing a document of max tx size length takes 1 second, and only then we charge gas on the parsing, then we need to limit it in a different way)

Reply:
parsing is currently 1000 times cheaper than verifying and we do charge on parsing before continuing to verify